src/terminal/adapter/SixelParser.cpp

```diff
@@ -0,0 +1,849 @@
+{
+    switch (conformanceLevel)
+    {
+    case 1: // Compatible with the VT125
+        return { 9, 20 };
+    default: // Compatible with the VT240 and VT340
+        return { 10, 20 };
+    }
+}
+
+size_t SixelParser::MaxColorsForLevel(const VTInt conformanceLevel) noexcept
+{
+    switch (conformanceLevel)
+    {
+    case 1: // Compatible with the monochrome VT125
+        return 2;
+    case 2: // Compatible with the 4-color VT240
+        return 4;
+    case 3: // Compatible with the 16-color VT340
+        return 16;
+    default: // Modern sixel apps often require 256 colors.
+        return MAX_COLORS;
+    }
+}
+
+SixelParser::SixelParser(AdaptDispatch& dispatcher, const StateMachine& stateMachine, const VTInt conformanceLevel) noexcept :
+    _dispatcher{ dispatcher },
+    _stateMachine{ stateMachine },
+    _conformanceLevel{ conformanceLevel },
+    _cellSize{ CellSizeForLevel(conformanceLevel) },
+    _maxColors{ MaxColorsForLevel(conformanceLevel) }
+{
+    if (_maxColors == 2)
+    {
+        // For monochrome support, we just initialize the first two color table
+        // entries with black and white.
+        til::at(_colorTable, 0) = til::color{ 0x00, 0x00, 0x00 };
+        til::at(_colorTable, 1) = til::color{ 0xCC, 0xCC, 0xCC };
+    }
+    else
+    {
+        // For everything else, we initialize the first 16 color entries with the
+        // VT340 palette, and use the XTerm extended colors for entries 17 to 255.
+        Microsoft::Console::Utils::InitializeVT340ColorTable(_colorTable);
+        Microsoft::Console::Utils::InitializeExtendedColorTable(_colorTable);
+    }
+}
+
+void SixelParser::SoftReset()
+{
+    // The VT240 is the only terminal known to reset colors with DECSTR.
+    // We only reset the first 16, since it only needs 4 of them anyway.
+    if (_conformanceLevel == 2)
+    {
+        Microsoft::Console::Utils::InitializeVT340ColorTable(_colorTable);
+        _updateTextColors();
+    }
+}
+
+void SixelParser::SetDisplayMode(const bool enabled) noexcept
+{
+    // The display mode determines whether images are clamped at the bottom of
+    // the screen (the set state), or scroll when they reach the bottom of the
+    // margin area (the reset state). Clamping was the only mode of operation
+    // supported prior to the VT340, so we don't allow the mode to be reset on
+    // levels 1 and 2.
+    if (_conformanceLevel >= 3)
+    {
+        _displayMode = enabled;
+    }
+}
+
+std::function<bool(wchar_t)> SixelParser::DefineImage(const VTInt macroParameter, const DispatchTypes::SixelBackground backgroundSelect, const VTParameter backgroundColor)
+{
+    if (_initTextBufferBoundaries())
+    {
+        _initRasterAttributes(macroParameter, backgroundSelect);
+        _initColorMap(backgroundColor);
+        _initImageBuffer();
+        _state = States::Normal;
+        _parameters.clear();
+        return [&](const auto ch) {

```

Comment:
I'm never sure about lambda capture, but I don't _think_ that's correct. I've just checked the [cppreference](https://en.cppreference.com/w/cpp/language/lambda#Lambda_capture) documentation, and according to them
> The current object (`*this`) can be implicitly captured if either capture default is present. If implicitly captured, it is always captured by reference, even if the capture default is `=`.

It then goes on to say
> The implicit capture of `*this` when the capture default is `=` is deprecated.

Which makes me think `[&]` is actually the preferred choice here, although they both work the same way. I'm happy to be corrected if I'm reading this wrong, but it might be worth getting a third opinion.

And I did try changing the code to `[=]` to see what affect that would have, and as far as I can see, the generated assembly is identical.

Reply:
Nope, you're totally right.

Now I need to remember which thing it was that burned me and made me think this way. 🙂